bundles/io/org.eclipse.smarthome.io.rest.core/src/main/java/org/eclipse/smarthome/io/rest/core/thing/ThingResource.java

```diff
@@ -515,6 +520,65 @@ public Response getConfigStatus(@HeaderParam(HttpHeaders.ACCEPT_LANGUAGE) String
         return Response.ok().entity(Collections.EMPTY_SET).build();
     }
 
+    @PUT
+    @Path("/{thingUID}/firmware/{firmwareVersion}")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @ApiOperation(value = "Update thing firmware.")
+    @ApiResponses(value = { @ApiResponse(code = 200, message = "OK"),
+            @ApiResponse(code = 400, message = "Unkown thing or firmware") })
+    public Response updateFirmware(
+            @HeaderParam(HttpHeaders.ACCEPT_LANGUAGE) @ApiParam(value = "language") String language,
+            @PathParam("thingUID") @ApiParam(value = "thing") String thingUID,
+            @PathParam("firmwareVersion") @ApiParam(value = "version") String firmwareVersion) throws IOException {
+        Thing thing = thingRegistry.get(new ThingUID(thingUID));
+        if (thing == null) {
+            return JSONResponse.createResponse(Status.BAD_REQUEST, null, "Unknown thingUID");
+        }
+
+        FirmwareUID firmwareUID = new FirmwareUID(thing.getThingTypeUID(), firmwareVersion);
+
+        try {
+            firmwareUpdateService.updateFirmware(thing.getUID(), firmwareUID, LocaleUtil.getLocale(language));
+        } catch (IllegalArgumentException | NullPointerException ex) {
+            return JSONResponse.createResponse(Status.BAD_REQUEST, null, "Unknown thing/firmware id");
+        }
+
+        return Response.status(Status.OK).build();
+    }
+
+    @GET
+    @Path("/{thingUID}/firmware/status")
+    @ApiOperation(value = "Gets thing's firmware status.")
+    @ApiResponses(value = { @ApiResponse(code = 200, message = "OK"),
+            @ApiResponse(code = 404, message = "Thing not found.") })
+    public Response getFirmwareStatus(@HeaderParam(HttpHeaders.ACCEPT_LANGUAGE) String language,
+            @PathParam("thingUID") @ApiParam(value = "thing") String thingUID) throws IOException {
+        ThingUID thingUIDObject = new ThingUID(thingUID);
+
+        FirmwareStatusInfo info = firmwareUpdateService.getFirmwareStatusInfo(thingUIDObject);
+        if (info == null) {
+            return JSONResponse.createErrorResponse(Status.NOT_FOUND, "Firmware status info not found");
+        }
+
+        return Response.ok().entity(buildFirmwareDTO(info)).build();
+    }
+
+    private FirmwareStatusDTO getThingFirmwareStatus(ThingUID thingUID) {
+        FirmwareStatusInfo info = firmwareUpdateService.getFirmwareStatusInfo(thingUID);
+        if (info != null) {
+            return buildFirmwareDTO(info);
+        }
+
+        return null;
+    }
+
+    private FirmwareStatusDTO buildFirmwareDTO(FirmwareStatusInfo info) {
+        String updatableFirmwareUID = info.getUpdatableFirmwareUID() == null ? ""
```

Comment:
I think the param name should contain something with version and not with uid....
 wouldnt it be more precise if we would return null instead of ""